[Security Solution] Add extra rules export tests

maximpn opened the pull request:
**Relates to:** https://github.com/elastic/security-team/issues/5339, https://github.com/elastic/kibana/pull/150097

## Summary

This PR covers rules exporting functionality by extra functional and e2e tests to make sure `execution_summary` and the other dynamic fields aren't exported. This problem was introduced during `8.7` dev cycle.

## Details

While working on this PR some changes were made to achieve the goal

- fixed unintentional exporting `execution_summary` field while exporting all rules
- a functional test was added to test exporting all rules after execution
- a functional test was added to test exporting a specific rule after execution
- an e2e test was added to test importing of just exported rules
- little refactoring to facilitate creation of enabled rules


### Checklist

Delete any items that are not applicable to this PR.

- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

maximpn committed:
delete execution_summary from export all rules function

maximpn committed:
cover rules export by functional tests

maximpn committed:
add an extra importing just exported rules e2e test

maximpn committed:
add waitForDelay helper function

maximpn commented on x-pack/test/detection_engine_api_integration/security_and_spaces/group1/export_rules.ts:
```diff
@@ -19,15 +19,28 @@ import {
 
   describe('export_rules', () => {
     describe('exporting rules', () => {
+      before(async () => {
```
`waitForRuleSuccessOrStatus()` fails without this chunk added

maximpn committed:
use waitForEventLogExecuteComplete instead of waitForDelay

maximpn committed:
use proper function invocation

maximpn committed:
fix expected exported rule object

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

maximpn committed:
fix rule creation

maximpn committed:
reuse expectManagementTableRules helper function

xcrzx commented on x-pack/test/detection_engine_api_integration/security_and_spaces/group1/export_rules.ts:
```diff
@@ -65,6 +77,58 @@ export default ({ getService }: FtrProviderContext): void => {
+
+        expect(bodySplitAndParsed).to.eql({
+          ...getSimpleRuleOutput(ruleId, true),
+          id: rule.id,
+          created_at: rule.created_at,
+          updated_at: rule.updated_at,
```
Can we use `expect.objectContaining` here to verify only relevant properties?

maximpn replied:
What's wrong in checking that we export specific fields only? The rule's export format shouldn't change easily. If some new fields appear it must be prominent.

xcrzx replied:
We have discussed the drawbacks of using snapshot tests multiple times, and it is worth summarizing them for clarity. Firstly, snapshot tests tend to capture irrelevant information, creating unnecessary noise in the output. Secondly, they hide the intent of the test, making it difficult for developers to determine precisely what matters just by looking at the output. Lastly, snapshot tests are often fragile and require frequent updates, leading to false positives.

To avoid these issues, it's important to make the intent of your test explicit. For instance, if your goal is to verify that dynamic fields do not get exported, it is best to focus on checking that a specific rule was successfully exported and that the output does not contain any dynamic fields. On the other hand, if you want to verify the correctness of the entire exported rule schema, it is better to create a separate test for that purpose. Mixing different cases in one test can lead to confusion and complicate the debugging process.

By explicitly stating the intent of your test and creating separate tests for different cases, you can improve the quality and reliability of your codebase. This approach reduces the likelihood of false positives and makes it easier for developers to understand the test results and debug issues quickly.

xcrzx reviewed (commented):
Thank you, @maximpn, for expanding the test coverage 👍 I have a few suggestions to improve the readability of the tests, but nothing is critical. In general, I think the changes look good to me!

maximpn commented on x-pack/test/detection_engine_api_integration/security_and_spaces/group1/export_rules.ts:
```diff
@@ -65,6 +77,58 @@ export default ({ getService }: FtrProviderContext): void => {
+        const rule = await createRule(supertest, log, getSimpleRule(ruleId, true));
+
+        await waitForRuleSuccessOrStatus(supertest, log, rule.id);
+        await waitForEventLogExecuteComplete(es, log, rule.id);
```
We wanna test exporting functionality in almost real use case which means rule's SO has all the fields including runtime fields like `execution_summary`. It takes time to write updates to SO and `execution_summary` isn't presented after `waitForRuleSuccessOrStatus()` returns.

banderror replied:
When we create an `execution_summary` we don't read any data for it from the event log. So it's not correct to rely on that in the tests. I assume the `await waitForEventLogExecuteComplete` call just takes time enough for something else to happen. Which is what exactly @maximpn?

maximpn replied:
@banderror you are right. `cy.wait()` may make the test flaky while `await waitForEventLogExecuteComplete()` looks more reliable to just wait some time. I'm gonna add an explicit `await waitForExecutionSummary()` here.

maximpn committed:
improve rule creation readability

maximpn committed:
improve test's naming

maximpn committed:
make sure rule has been executed before exporting

maximpn committed:
remove unnecessary wait command

maximpn committed:
add a comment to transformRuleToExportableFormat

maximpn committed:
move constants to files their belong to

maximpn committed:
fix typing errors

maximpn committed:
move rule export tests related logic to the test itself

maximpn commented on x-pack/plugins/security_solution/cypress/e2e/detection_rules/export_rule.cy.ts:
```diff
@@ -45,18 +53,33 @@ describe('Export rules', () => {
   });
 
-  it('shows a modal saying that no rules can be exported if all the selected rules are prebuilt', function () {
+  it('imports an exported custom rule', () => {
```
Good suggestion, I've renamed the test and added a construction to wait for rule execution.

maximpn commented on x-pack/plugins/security_solution/cypress/e2e/detection_rules/export_rule.cy.ts:
```diff
@@ -45,18 +53,33 @@ describe('Export rules', () => {
     cy.intercept('POST', '/api/detection_engine/rules/_bulk_action').as('bulk_action');
     visitWithoutDateRange(DETECTIONS_RULE_MANAGEMENT_URL);
-    createCustomRule(getNewRule()).as('ruleResponse');
+    createCustomRule({ ...getNewRule(), name: 'Rule to export' }, { enabled: true }).as(
```
Removed it here and added an enabled rule creation in a specific test.

maximpn committed:
add an explanation comment

maximpn commented on x-pack/plugins/security_solution/cypress/objects/rule.ts:
```diff
@@ -531,7 +530,7 @@ export const expectedExportedRule = (ruleResponse: Cypress.Response<RuleResponse
     tags,
     interval: '100m',
-    enabled: false,
+    enabled: true,
```
done

maximpn committed:
add getNewSavedQueryRule helper

maximpn committed:
fix rule id usage

maximpn committed:
move id from createCustomRuleEnabled to getNewRule

maximpn committed:
fix enabled rule creation

maximpn committed:
simplify getNewSavedQueryRule

maximpn commented:
After deep investigation I've found a reason why there is a need in some delay after waiting for rule's status in functional tests. I have the code like this

```
await waitForRuleSuccessOrStatus(supertest, log, rule.id);
await waitForEventLogExecuteComplete(es, log, rule.id);

const { body } = await supertest
  .post(`${DETECTION_ENGINE_RULES_URL}/_export`)
  .set('kbn-xsrf', 'true')
  .send()
  .expect(200)
  .parse(binaryToString);

const bodySplitAndParsed = JSON.parse(body.toString().split(/\n/)[0]);

expect(bodySplitAndParsed).to.eql({...});
```

I needed it to make sure `execution_summary` is presented in the rule's SO but it doesn't get exported. The problem is that `execution_summary` field can be missed after `await waitForRuleSuccessOrStatus()` resolves.

If you look inside `waitForRuleSuccessOrStatus()` it becomes clear the status is read from `execution_summary` field and one expects to see this field when the promise returned by the function resolves.

It's not a secret we support two distinct rule identifiers an SO `id` and a `ruleId` saved in `alert.attributes.params.ruleId` SO field. `waitForRuleSuccessOrStatus()` relies on read rule endpoint which invokes [`readRules` function](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/crud/read_rules.ts#L34) and starting from here we have two branches as `readRules` accepts `id` and `ruleId`

- `id` provided, `rulesClient.resolve() -> SOclient.resolve() -> ES Get API`
- `ruleId` provided `findRules() -> rulesClient.find() -> SOClient.find() -> ES Search API`

The difference between them is whether outdated data can be returned. ES Get API always returns fresh data while ES Search API can return outdated data. Since `waitForRuleSuccessOrStatus()` relies on ES Get API while rule exporting functionality relies on ES Search API an exported rule can have `execution_summary` field missing while our goal is to make the test 100% reliable.

There are some ways to resolve it
- use only `ruleId` in functional tests (or all tests)
- let `waitForRuleSuccessOrStatus()` accept `ruleId`
- wait for `execution_summary` via find rules endpoint

@banderror @xcrzx @spong @jpdjere please read my comments and let's think on the solution. You may have some good ideas to resolve it different from mine.

banderror commented:
Thanks for the very clear explanation of the problem @maximpn 🙏 My vote would go for `wait for execution_summary via find rules endpoint` because that would tell that it relies on the ES `_search` API in the most explicit way -- just like the export relies on it.

The other options:

- `let waitForRuleSuccessOrStatus() accept ruleId` works too but is more implicit, because using the `_search` API when `rule_id` is provided instead of `id` is an implementation detail
- `use only ruleId in functional tests (or all tests)` we should be free to use `id` where we can or have to and where we don't have the same problem, e.g. in tests for rule CRUD or bulk actions -- `id` is a more reliable and precise identified

maximpn committed:
allow to wait for rule status by ruleId

maximpn committed:
remove unused constant

maximpn committed:
fix typing errors

xcrzx commented on x-pack/test/detection_engine_api_integration/utils/wait_for_rule_status.ts:
```diff
@@ -7,33 +7,51 @@
+  afterDate?: Date;
+}
+
+interface WaitForRuleStatusWithId extends WaitForRuleStatusBaseParams {
+  id: string;
+}
+
+interface WaitForRuleStatusWithRuleId extends WaitForRuleStatusBaseParams {
+  ruleId: string;
+}
```
If I get your idea correctly, for these types to become discriminated unions, you need to change them to:

```ts
interface WaitForRuleStatusWithId extends WaitForRuleStatusBaseParams {
  id: string;
  ruleId?: undefined;
}

interface WaitForRuleStatusWithRuleId extends WaitForRuleStatusBaseParams {
  id?: undefined;
  ruleId: string;
}
```

That way, either `id` or `ruleId` will have to be provided to `waitForRuleStatus`, but not both of them at the same time.

maximpn commented on x-pack/test/detection_engine_api_integration/security_and_spaces/group1/export_rules.ts:
```diff
@@ -65,6 +77,58 @@ export default ({ getService }: FtrProviderContext): void => {
+        const rule = await createRule(supertest, log, getSimpleRule(ruleId, true));
+
+        await waitForRuleSuccessOrStatus(supertest, log, rule.id);
+        await waitForEventLogExecuteComplete(es, log, rule.id);
```
See this comment https://github.com/elastic/kibana/pull/150553#issuecomment-1456541085

maximpn commented on x-pack/test/detection_engine_api_integration/security_and_spaces/group1/export_rules.ts:
```diff
@@ -65,6 +77,58 @@ export default ({ getService }: FtrProviderContext): void => {
+
+        const bodySplitAndParsed = JSON.parse(body.toString().split(/\n/)[0]);
+
+        expect(bodySplitAndParsed).to.eql({
+          ...getSimpleRuleOutput(ruleId, true),
+          id: rule.id,
+          created_at: rule.created_at,
+          updated_at: rule.updated_at,
+        });
```
The same as above. It looks right to check the whole rule's object.

marshallmain commented on x-pack/plugins/security_solution/cypress/objects/rule.ts:
```diff
@@ -359,9 +363,10 @@ export const getNewThresholdRule = (): ThresholdRule => ({
 });
 
-export const getNewTermsRule = (): NewTermsRule => ({
+export const getNewTermsRule = (rewrites?: Partial<NewTermsRule>): NewTermsRule => ({
```
Similar to below, the [other PR](https://github.com/elastic/kibana/pull/151520) will make broad changes to the rule mocks, so we'll get significant conflicts here. Once we have the new schemas in place for these functions, I think it will also be easier to apply rewrites to the mock data using a pattern like `{ ...getNewTermsRule(), rule_id: 'new_rule_id'}`.

maximpn replied:
@xcrzx @marshallmain's suggestion conflicts with your https://github.com/elastic/kibana/pull/150553#discussion_r1115537401. I personally prefer provide rewrites to mock creators like this `getNewTermsRule({ rule_id: 'new_rule_id' })` but it will be good to find common ground.

xcrzx replied:
I would prefer passing overrides as an argument to the mock builder as it makes the code appear more organized and provides greater flexibility. For instance, we could effortlessly implement a deep merge of the mocked structure with the provided override. 

For example, overriding nested properties become tiresome quickly, as shown in the following code snippet:

```ts
const rule = {
  ...getNewTermsRule(),
  rule_id: 'new_rule_id',
  runsEvery: {
    interval: '1',
    ...getNewTermsRule().runsEvery,
  },
};
```

If we implement deep merge, the readability could be greatly improved:

```ts
const rule = getNewTermsRule({
  rule_id: 'new_rule_id',
  runsEvery: {
    interval: '1',
  },
});
```

maximpn committed:
match test names with contents

maximpn committed:
simplify waitForRuleStatus logic

maximpn committed:
fix rule's identifier discriminated union

maximpn commented:
@xcrzx I've addressed all your comments. Regarding https://github.com/elastic/kibana/pull/150553#discussion_r1128073924 it looks like we've got misunderstanding here. I've changed test names to match the test internals check it here https://github.com/elastic/kibana/pull/150553/commits/6f1144e6283c35b4689e1e5a2ef57a656ed2faf0 where I

maximpn commented on x-pack/plugins/security_solution/cypress/tasks/api_calls/rules.ts:
```diff
@@ -39,8 +39,7 @@ export const createMachineLearningRule = (rule: MachineLearningRule, ruleId = 'm
 export const createCustomRule = (
-  rule: CustomRule,
-  ruleId = 'rule_testing'
+  rule: CustomRule
```
Sure, np. I'll update the PR after yours merged.

maximpn commented:
After appearing of the huge number of the conflicts it's simpler to split this PR into two parts.
